src/libraries/Microsoft.Extensions.Hosting.WindowsServices/src/WindowsServiceLifetime.cs

```diff
@@ -92,14 +93,14 @@ private void Run()
             {
                 _delayStart.TrySetException(ex);
             }
+            _serviceStopped.TrySetResult(null);
         }
 
         public Task StopAsync(CancellationToken cancellationToken)
         {
-            // Avoid deadlock where host waits for StopAsync before firing ApplicationStopped,
-            // and Stop waits for ApplicationStopped.
+            // Stop will cause the ServiceBase.Run method to complete and return, which completes _serviceStopped.
             Task.Run(Stop, CancellationToken.None);
```

Comment:
This stop is redundant in the case that we received a stop signal from SCM.  It's here for the case where the app is deciding to stop the service itself.  We should try to avoid having this run concurrently with SCM's DeferredStop.

Reply:
One thing that would be ordered is the call of OnStop.  We could set a flag at that point and honor here to avoid calling Stop if OnStop has already been called.